hi-ml-histopathology/src/histopathology/datamodules/base_module.py

```diff
@@ -150,40 +201,139 @@ def _load_dataset(self, tiles_dataset: TilesDataset, stage: str, shuffle: bool)
+            Defaults to ``False``.
+        pad_full: pad image to the size evenly divisible by tile_size
+            Defaults to ``False``.
+        background_val: the background constant (e.g. 255 for white background)
+            Defaults to ``255``.
+        filter_mode: mode must be in ["min", "max", "random"]. If total number of tiles is more than tile_size,
+            then sort by intensity sum, and take the smallest (for min), largest (for max) or random (for random) subset
+            Defaults to ``min`` (which assumes background is high value)
+    """
 
-    def test_dataloader(self) -> DataLoader:
-        return self._get_dataloader(self.test_dataset, 'test', shuffle=True, **self.dataloader_kwargs)
+    def __init__(
+        self,
+        level: Optional[int] = 0,
+        tile_count: Optional[int] = None,
+        tile_size: Optional[int] = 224,
+        step: Optional[int] = None,
+        random_offset: Optional[bool] = False,
+        pad_full: Optional[bool] = False,
+        background_val: Optional[int] = 255,
+        filter_mode: Optional[str] = "min",
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.level = level
+        self.tile_count = tile_count
+        self.tile_size = tile_size
+        self.step = step
+        self.random_offset = random_offset
+        self.pad_full = pad_full
+        self.background_val = background_val
+        self.filter_mode = filter_mode
+
+    def prepare_data(self) -> None:
+        if self.precache_location != CacheLocation.NONE:
+            self._load_dataset(self.train_dataset, stage="train")
+            self._load_dataset(self.val_dataset, stage="val")
+            self._load_dataset(self.test_dataset, stage="test")
+
+    def _load_dataset(self, slides_dataset: SlidesDataset, stage: str) -> Dataset:
+        dataset_pickle_path = self._dataset_pickle_path(stage)
+
+        if dataset_pickle_path and dataset_pickle_path.is_file():
+            if self.precache_location == CacheLocation.CPU:
+                memory_location = torch.device("cpu")
+                print(f"Loading dataset from {dataset_pickle_path} into {memory_location}")
+            else:
+                # by default torch.load will reload on the same device it was saved from
+                memory_location = None  # type: ignore
+
+            with dataset_pickle_path.open("rb") as f:
+                return torch.load(f, map_location=memory_location)
+
+        base_transform = Compose(
+            [
+                LoadImaged(
+                    keys=slides_dataset.IMAGE_COLUMN,
+                    reader=WSIReader,
+                    backend="cucim",
+                    dtype=np.uint8,
+                    level=self.level,
+                    image_only=True,
+                ),
+                TileOnGridd(
+                    keys=PandaDataset.IMAGE_COLUMN,
```

Comment:
this should be a generic class, we should remove any hard-coded dependency from specific datasets, we can pass this variable to the constructor if it's not available